How many late days are you using for this assignment? 
0

In [147]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder, Normalizer, MinMaxScaler

# Part 1: Load the dataset

In [148]:
# Load the given datasets
kidney_cat = pd.read_csv('chronic_kidney_disease_categorical.csv')
kidney_num = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print(kidney_cat)
print(kidney_num)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: This dataset, collected from a hospital over a two month timeframe, contains medical data from patients with and without kidney disease. This dataset can be used to train ML models to predict kidney disease.

The features of this dataset are listed below:
    1.Age(numerical)
  	  	    age in years
 	2.Blood Pressure(numerical)
	       	bp in mm/Hg
 	3.Specific Gravity(categorical/nominal)
	  	sg - (1.005,1.010,1.015,1.020,1.025)
 	4.Albumin(categorical/nominal)
		al - (0,1,2,3,4,5)
 	5.Sugar(categorical/nominal)
		su - (0,1,2,3,4,5)
 	6.Red Blood Cells(categorical/nominal)
		rbc - (normal,abnormal)
 	7.Pus Cell (categorical/nominal)
		pc - (normal,abnormal)
 	8.Pus Cell clumps(categorical/nominal)
		pcc - (present,notpresent)
 	9.Bacteria(categorical/nominal)
		ba  - (present,notpresent)
 	10.Blood Glucose Random(numerical)		
		bgr in mgs/dl
 	11.Blood Urea(numerical)	
		bu in mgs/dl
 	12.Serum Creatinine(numerical)	
		sc in mgs/dl
 	13.Sodium(numerical)
		sod in mEq/L
 	14.Potassium(numerical)	
		pot in mEq/L
 	15.Hemoglobin(numerical)
		hemo in gms
 	16.Packed  Cell Volume(numerical)
 	17.White Blood Cell Count(numerical)
		wc in cells/cumm
 	18.Red Blood Cell Count(numerical)	
		rc in millions/cmm
 	19.Hypertension(categorical/nominal)	
		htn - (yes,no)
 	20.Diabetes Mellitus(categorical/nominal)	
		dm - (yes,no)
 	21.Coronary Artery Disease(categorical/nominal)
		cad - (yes,no)
 	22.Appetite(categorical/nominal)	
		appet - (good,poor)
 	23.Pedal Edema(categorical/nominal)
		pe - (yes,no)	
 	24.Anemia(categorical/nominal)
		ane - (yes,no)
 	25.Class (categorical/nominal)		
		class - (ckd,notckd)

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [149]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(sum(kidney_num.duplicated()))

# Drop these duplicate rows
kidney_num = kidney_num.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(sum(kidney_cat.duplicated()))
kidney_cat = kidney_cat.drop_duplicates()


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: The unique IDs are likely not all present in both datasets, as they are not both the same size. In the combined table, rows that are missing will be filled with several NA values.

In [150]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
kidney_all = pd.merge(left=kidney_cat, right=kidney_num, on='unique_id')
#Print the combined dataset
print(kidney_all)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [151]:
# Calculate the percentage of rows that contain atleast one missing value
pct_na = kidney_all.isna().any(axis=1).sum() / kidney_all.shape[0]
# Print %
print(pct_na)
# Drop these rows from the dataset
kidney_all = kidney_all.dropna()
# Print the Dataset
print(kidney_all)

0.6076923076923076
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  normal  notpresent  notpresent   no   no   

     cad  ...     bp

## Part 3.4: Sort the dataset according to the Labels

In [152]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
kidney_all = kidney_all.sort_values(by='Target')
kidney_all = kidney_all.reset_index()
# Print the dataset
print(kidney_all)

     index  unique_id   al   su       rbc        pc         pcc          ba  \
0      118     534245  2.0  0.0    normal  abnormal     present  notpresent   
1      190     349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent   
2      337     343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent   
3      272     102677  3.0  0.0    normal  abnormal  notpresent  notpresent   
4      329     514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent   
..     ...        ...  ...  ...       ...       ...         ...         ...   
148    127     489160  0.0  0.0    normal    normal  notpresent  notpresent   
149    123     530138  0.0  0.0    normal    normal  notpresent  notpresent   
150    121     261137  0.0  0.0    normal    normal  notpresent  notpresent   
151    111     322882  0.0  0.0    normal    normal  notpresent  notpresent   
152    388     273038  0.0  0.0    normal    normal  notpresent  notpresent   

     htn   dm  ...    bp    bgr     bu    sc    sod

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [153]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

encoder = OneHotEncoder(handle_unknown='ignore')
encoder.fit(kidney_all[categorical_columns])

encoded = encoder.transform(kidney_all[categorical_columns])
encoded_df = pd.DataFrame(encoded.toarray(), columns=encoder.get_feature_names_out(categorical_columns))

kidney_all = kidney_all.drop(columns=categorical_columns)
kidney_all = pd.concat([kidney_all.reset_index(drop=True), encoded_df.reset_index(drop=True)], axis=1)
# Print the dataset
print(kidney_all)

     index  unique_id   al   su   age    bp    bgr     bu    sc    sod  ...  \
0      118     534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  ...   
1      190     349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  ...   
2      337     343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  ...   
3      272     102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  ...   
4      329     514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  ...   
..     ...        ...  ...  ...   ...   ...    ...    ...   ...    ...  ...   
148    127     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  ...   
149    123     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  ...   
150    121     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  ...   
151    111     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  ...   
152    388     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer:  The binary variables may be treated as ordinal when they are not, and one hot encoding is a way to avoid this error. One hot encoding is also excellent for making data compatible with certain algorithms and for variables with higher cardinality this method with greatly increase the size of the dataset, possibly adding unnecessary complexity to the model.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [154]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

means = kidney_all[numerical_columns].mean(axis=0)
numerical_deviations = kidney_all[numerical_columns].std(axis=0)

upper_limits = means + 3 * numerical_deviations
lower_limits = means - 3 * numerical_deviations

for column in numerical_columns:
    kidney_all = kidney_all[(kidney_all[column] >= lower_limits[column]) & (kidney_all[column] <= upper_limits[column])]

kidney_all = pd.DataFrame(kidney_all)

# Optionally, to reset index if needed
kidney_all.reset_index(drop=True, inplace=True)
# Print the dataset
print(kidney_all)

     index  unique_id   al   su   age    bp    bgr     bu   sc    sod  ...  \
0      337     343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  ...   
1      329     514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  ...   
2      163     124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  ...   
3      314     980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  ...   
4      277     615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  ...   
..     ...        ...  ...  ...   ...   ...    ...    ...  ...    ...  ...   
129    127     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  ...   
130    123     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  ...   
131    121     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  ...   
132    111     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  ...   
133    388     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  an

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [155]:
# Normalize the all Numerical Attributes in the dataset.
normalizer = MinMaxScaler()

normalized_data = normalizer.fit_transform(kidney_all[numerical_columns])
normalized_numeric_df = pd.DataFrame(normalized_data, columns=numerical_columns)

kidney_all[numerical_columns] = normalized_numeric_df

# Print the dataset
print(kidney_all)

     index  unique_id   al   su       age    bp       bgr        bu        sc  \
0      337     343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1      329     514721  2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778   
2      163     124304  3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951   
3      314     980291  3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494   
4      277     615697  4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074   
..     ...        ...  ...  ...       ...   ...       ...       ...       ...   
129    127     489160  0.0  0.0  0.297297  0.00  0.038627  0.261438  0.012346   
130    123     530138  0.0  0.0  0.229730  0.50  0.175966  0.156863  0.086420   
131    121     261137  0.0  0.0  0.391892  0.00  0.150215  0.189542  0.012346   
132    111     322882  0.0  0.0  0.364865  0.50  0.248927  0.183007  0.024691   
133    388     273038  0.0  0.0  0.554054  0.00  0.201717  0.078431  0.098765   

          sod  ...  cad_no 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: columns 'unique_id' and 'index' are unnecessary for modeling. These columns are both meant to identify particular instances of data, and they will add nothing of value to our model.

In [156]:
#Remove that column
kidney_all = kidney_all.drop(['unique_id', 'index'], axis=1)
# Print the dataset
print(kidney_all)

      al   su       age    bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1    2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
2    3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
3    3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
4    4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
..   ...  ...       ...   ...       ...       ...       ...       ...   
129  0.0  0.0  0.297297  0.00  0.038627  0.261438  0.012346  0.833333   
130  0.0  0.0  0.229730  0.50  0.175966  0.156863  0.086420  0.833333   
131  0.0  0.0  0.391892  0.00  0.150215  0.189542  0.012346  0.500000   
132  0.0  0.0  0.364865  0.50  0.248927  0.183007  0.024691  0.500000   
133  0.0  0.0  0.554054  0.00  0.201717  0.078431  0.098765  0.600000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.657143  0.172131  ...     0.0      1

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [157]:
# Export the dataframe to a new csv file
kidney_all.to_csv('kidney_all.csv')

# Export the dataframe to a new json file
kidney_all.to_json('kidney_all.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [158]:
print(kidney_all[:15])

     al   su       age    bp       bgr        bu        sc       sod  \
0   2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1   2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
2   3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
3   3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
4   4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
5   3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
6   1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
7   4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
8   4.0  0.0  0.567568  0.50  0.107296  1.000000  0.901235  0.533333   
9   4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
10  3.0  2.0  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
11  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
12  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.56

Paste here:      al   su       age    bp       bgr        bu        sc       sod  \
0   2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1   2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
2   3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
3   3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
4   4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
5   3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
6   1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
7   4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
8   4.0  0.0  0.567568  0.50  0.107296  1.000000  0.901235  0.533333   
9   4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
10  3.0  2.0  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
11  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
12  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
13  4.0  1.0  0.675676  0.25  0.600858  0.104575  0.160494  0.533333   
14  4.0  0.0  0.567568  0.50  0.270386  0.843137  1.000000  0.400000   

         pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0   0.657143  0.172131  ...     0.0      1.0         0.0         1.0    0.0   
1   0.542857  0.286885  ...     1.0      0.0         1.0         0.0    1.0   
2   0.314286  0.565574  ...     0.0      1.0         1.0         0.0    0.0   
3   0.514286  0.573770  ...     1.0      0.0         1.0         0.0    1.0   
4   0.571429  0.352459  ...     1.0      0.0         0.0         1.0    1.0   
5   0.571429  0.434426  ...     1.0      0.0         1.0         0.0    0.0   
6   0.314286  0.344262  ...     1.0      0.0         1.0         0.0    1.0   
7   0.485714  0.188525  ...     1.0      0.0         1.0         0.0    0.0   
8   0.257143  0.344262  ...     1.0      0.0         1.0         0.0    1.0   
9   0.285714  0.311475  ...     0.0      1.0         1.0         0.0    0.0   
10  0.000000  0.295082  ...     0.0      1.0         0.0         1.0    1.0   
11  0.171429  0.221311  ...     1.0      0.0         1.0         0.0    1.0   
12  0.714286  0.319672  ...     1.0      0.0         0.0         1.0    0.0   
13  0.257143  0.860656  ...     1.0      0.0         1.0         0.0    1.0   
14  0.742857  0.385246  ...     1.0      0.0         1.0         0.0    0.0   

    pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd  
0      1.0     0.0      1.0         1.0            0.0  
1      0.0     1.0      0.0         1.0            0.0  
2      1.0     1.0      0.0         1.0            0.0  
3      0.0     1.0      0.0         1.0            0.0  
4      0.0     1.0      0.0         1.0            0.0  
5      1.0     1.0      0.0         1.0            0.0  
6      0.0     1.0      0.0         1.0            0.0  
7      1.0     1.0      0.0         1.0            0.0  
8      0.0     0.0      1.0         1.0            0.0  
9      1.0     0.0      1.0         1.0            0.0  
10     0.0     1.0      0.0         1.0            0.0  
11     0.0     0.0      1.0         1.0            0.0  
12     1.0     1.0      0.0         1.0            0.0  
13     0.0     1.0      0.0         1.0            0.0  
14     1.0     1.0      0.0         1.0            0.0  



| al   | su   | age      | bp    | bgr      | bu       | sc       | sod      | pot      | hemo      | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|----------|-------|----------|----------|----------|----------|----------|-----------|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 2.0  | 0.0  | 0.743243 | 0.50  | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.657143 | 0.172131  | 0.0    | 1.0     | 0.0        | 1.0         | 0.0   | 1.0    | 0.0    | 1.0     | 1.0         | 0.0            |
| 2.0  | 0.0  | 0.675676 | 0.75  | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.542857 | 0.286885  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 3.0  | 4.0  | 0.851351 | 0.25  | 0.832618 | 0.503268 | 0.283951 | 0.333333 | 0.314286 | 0.565574  | 0.0    | 1.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 3.0  | 0.0  | 0.756757 | 0.25  | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.514286 | 0.573770  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 0.0  | 0.000000 | 0.00  | 0.103004 | 0.372549 | 0.074074 | 0.500000 | 0.571429 | 0.352459  | 1.0    | 0.0     | 0.0        | 1.0         | 1.0   | 0.0    | 0.0    | 0.0     | 1.0         | 0.0            |
| 3.0  | 1.0  | 0.662162 | 0.50  | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.571429 | 0.434426  | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 1.0  | 0.0  | 0.540541 | 0.00  | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.314286 | 0.344262  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 2.0  | 0.783784 | 1.00  | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.485714 | 0.188525  | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 0.0  | 0.567568 | 0.50  | 0.107296 | 1.000000 | 0.901235 | 0.533333 | 0.257143 | 0.344262  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 0.0    | 1.0     | 1.0         | 0.0            |
| 4.0  | 3.0  | 0.851351 | 0.25  | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.285714 | 0.311475  | 0.0    | 1.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 3.0  | 2.0  | 0.905405 | 1.00  | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | 0.295082  | 0.0    | 1.0     | 0.0        | 1.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 0.0  | 0.202703 | 0.75  | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.171429 | 0.221311  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 0.0    | 1.0     | 1.0         | 0.0            |
| 4.0  | 1.0  | 0.783784 | 0.00  | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.714286 | 0.319672  | 1.0    | 0.0     | 0.0        | 1.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 1.0  | 0.675676 | 0.25  | 0.600858 | 0.104575 | 0.160494 | 0.533333 | 0.257143 | 0.860656  | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0         | 0.0            |
| 4.0  | 0.0  | 0.567568 | 0.50  | 0.270386 | 0.843137 | 1.000000 | 0.400000 | 0.742857 | 0.385246  | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0         | 0.0            |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [159]:
# Code to rename all the columns in the dataset
# Assuming kidney_all is your DataFrame

# Define the mapping for renaming columns based on the provided column names
column_mapping = {
    'al': 'albumin',
    'su': 'sugar',
    'age': 'age',
    'bp': 'blood pressure',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'target chronic kidney disease',
    'Target_notckd': 'target not chronic kidney disease'
}

# Rename the columns in the DataFrame
kidney_all.rename(columns=column_mapping, inplace=True)

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [160]:
# Converted SQL to Pandas code
# Replace with your actual DataFrame name
your_dataframe = pd.read_csv('chronic_kidney_disease_categorical.csv')  # Your DataFrame initialization goes here

# Group by 'Target' and count occurrences
result = your_dataframe.groupby('Target').size().reset_index(name='count')

# Alternatively, you can use:
# result = your_dataframe.groupby('Target')['Target'].count().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd    264
1   ckd\t      2
2  notckd    164


The code counts the total number of instances with and without chronic kidney disease.